### ISSUE TITLE
Zoltan2 graph model method getEdgeList() seems to return the wrong weights

### ISSUE THREAD

**Issue by vjleung** [trilinos/Trilinos#316]
@trilinos/zoltan2 
Graph model method getEdgeList() seems to return the wrong weights when called from src/util/Zoltan2_Metric.hpp, called from src/util/Zoltan2_EvaluatePartition.hpp, called from test/helper/Zoltan2_EvaluatePartitionFactory.hpp, called from test/driver/test_driver.cpp with input file zoltan/test/ch_ewgt/ewgt.graph.

**Comment by kddevin** [comment]
Can you tell where the edge-weight problem occurs?  That is, are the correct or wrong in the driver's data structures?  in the values returned by the Adapter?  in the GraphModel?

**Comment by vjleung** [comment]
I do not know where they first go wrong, but on return from the call to getEdgeList(), from the GraphModel,  in globalWeightedCutsByPart(), the weights do not agree with the input file.  I printed them out right after that call.  They are not even symmetric.

**Comment by vjleung** [comment]
I am not sure what happened, but I am now getting different answers for chacoGraphMetricsTest.xml.

**Comment by vjleung** [comment]
I have traced the edge-weight problem back to the driver's Adapter, before the Graph Model.  I pushed some code to print out the edge-weights from the Adapter in the driver.

________________________________

**Comment by vjleung** [comment]
This will summarize my current understanding of the situation including some previous info and some new info.

I have traced the problem through the graph model in evaluate partition and the xpetra crs graph adapter in evaluate partition, the evaluate partition factory, and the test driver.

I printed out the weights immediately after the graph model is constructed and they do not match the input file, zoltan/test/ch_ewgt/ewgt.graph.  For example, the weights between vertices 6 and 7 are 1 in both directions in the input file but 1 and 5.5 in the two directions from the graph model.  Note that these are vertices 7 and 8 in the input file since it is 1 based while everything else is 0 based.  There are 10 weights in the input file that are not 1: 2 are .9, 4 are 4.5, and 4 are 5.5.  There are 10 weights in the graph model that are not 1:  2 are .9, 3 are 4.5, and 5 are 5.5.

I printed out the weights from the xpetra crs graph adapter immediately before it is passed to the graph model constructor in evaluate partition and they do not match either the input file or the graph model constructed.  While it has the above error, it also has an error between vertices 1 and 2.  These weights are .9 in the input file and graph model but 1 in the xpetra crs graph adapter.  There are 10 weights in the xpetra crs graph adapter that are not 1:  1 is .9, 4 are 4.5, and 5 are 5.5.

The same for the xpetra crs graph adapter immediate before it is passed to evaluate partition in the evaluate partition factory.

The same for the xpetra crs graph adapter immediate before it is passed to the evaluate partition factory in the test driver.  I pushed some code that does this.

The same for the xpetra crs graph adapter immediately before it is passed to the problem factory in the test driver.

Karen said the problem should not be in the reading of the input since that is done with the chaco reader.

So somewhere between reading the input and passing the xpetra crs graph adapter to the problem factory in the test driver, the weights are mangled.

**Comment by kddevin** [comment]
Vitus:  Are the edge weights incorrect also when running with only one MPI rank?  or only in parallel?  How about in parallel without distributing the input data?  It is possible that the weights are read correctly but not distributed correctly.
Karen

**Comment by vjleung** [comment]
Yes, I am checking now

Sent from my iPhone

On May 5, 2016, at 9:17 AM, K Devine <notifications@github.com<mailto:notifications@github.com>> wrote:


Vitus: Are the edge weights incorrect also when running with only one MPI rank? or only in parallel? How about in parallel without distributing the input data? It is possible that the weights are read correctly but not distributed correctly.
Karen

-

**Comment by vjleung** [comment]
I just got the same weights with one processor.

Sent from my iPhone

On May 5, 2016, at 9:17 AM, K Devine <notifications@github.com<mailto:notifications@github.com>> wrote:


Vitus: Are the edge weights incorrect also when running with only one MPI rank? or only in parallel? How about in parallel without distributing the input data? It is possible that the weights are read correctly but not distributed correctly.
Karen

-

**Comment by kddevin** [comment]
@MicheldeMessieres observed that the edge IDs (neighbor vertices) read from the Chaco files are not sorted, but when they are printed from the adapter, they are sorted.  (Good catch, Michel!)

Indeed, I verified that Tpetra's fillComplete sorts the edge IDs (column indices), so the output of UserInputForTests has the new order.  Unfortunately, we do not also re-sort the edge weights read from Chaco using this new order.  So the deficiency is our assumption that Tpetra retains the same edge-list ordering that we passed to it.

We'll need to modify UserInputForTests' code that reads Chaco files to do the reordering of the weights.  

I think the simplest solution is to (1) limit UserInputForTests to the case of one weight per edge (which is all our partitioners handle, anyway), and (2) store the edge weights as nonzero values in the Tpetra matrix that UserInputForTests uses as its data structure, rather than in a separate array.  This approach has the advantage that the edge weights get distributed "for free" when we redistribute the matrix.  It has the disadvantage that we remove the generality of multiple weights per edge. 

Opinions?

If folks are OK with this plan, Michel, Vitus and I can work on it this week.

@krcb @trilinos/zoltan2
